File: freeswitch_lua/route/call_limit.lua
--[[call limit
api command:
lua call_limit.lua @pattern @key  @limits
  @key: Limit the unique name of the calling object, such as phone number, channel name, gateway name, etc.
  @pattern: 
     count-of-day: Limit the number of calls per day.
	 count-of-month: Limit the number of calls per month.
  @limits: Limited number.	 

]]
if session ~= nil then
    session:setAutoHangup(false)
end	

local rootdir = freeswitch.getGlobalVariable("base_dir")
loadfile(rootdir.."/scripts/include/functions.lua")()
local api =  freeswitch.API();

function reset_count_of_day(key)
    if key == nil or key == "all" then
		local keys = string_spilt(api:executeString("db list/call_limit_last_count_of_day/"),",")
		for k,v in ipairs(keys) do
            api:executeString("db delete/call_limit_last_count_of_day/"..v)		
		end
		local keys = string_spilt(api:executeString("db list/call_limit_value_count_of_day/"),",")
		for k,v in ipairs(keys) do
            api:executeString("db delete/call_limit_value_count_of_day/"..v)		
		end	
    else
        api:executeString("db delete/call_limit_value_count_of_day/"..key)	
		api:executeString("db delete/call_limit_last_count_of_day/"..key)	
	end
end

function reset_count_of_month(key)
    if key == nil or key == "all" then
		local keys = string_spilt(api:executeString("db list/call_limit_last_count_of_month/"),",")
		for k,v in ipairs(keys) do
            api:executeString("db delete/call_limit_last_count_of_month/"..v)		
		end
		local keys = string_spilt(api:executeString("db list/call_limit_value_count_of_month/"),",")
		for k,v in ipairs(keys) do
            api:executeString("db delete/call_limit_value_count_of_month/"..v)		
		end	
    else
        api:executeString("db delete/call_limit_value_count_of_month/"..key)	
		api:executeString("db delete/call_limit_last_count_of_month/"..key)	
	end
end

if argv[1] == 'reset' then
	if argv[2] == 'all' or argv[2] == nil then
	    reset_count_of_day("all")
	    reset_count_of_month("all")	
	elseif argv[2] == 'count-of-day' then
	    if argv[3] == nil or argv[3] == "all" then
		    reset_count_of_day("all")
		else
            reset_count_of_day(argv[3])
        end		
    elseif argv[2] == 'count-of-month' then      
	    if argv[3] == nil or argv[3] == "all" then
		    reset_count_of_month("all")
		else
            reset_count_of_month(argv[3])
        end			 
	end
	stream:write("+OK\ncall limit reset. ")
    return
end

if argv[1] == nil then
	freeswitch.consoleLog("ERR","must argv[1] pattern!\n")
	stream:write("false")
	return 
end 
local pattern = argv[1]

if argv[2] == nil then
	freeswitch.consoleLog("ERR","must argv[2] key!\n")
	stream:write("false")
	return
end 
local key= argv[2]


if argv[3] == nil then
	freeswitch.consoleLog("ERR","must argv[1] limits!\n")
	stream:write("false")
	return 
end 
local limits = tonumber(argv[3])



if pattern == 'count-of-day' then
    local last = nil
	if api:executeString("db exists/call_limit_last_count_of_day/"..key) == "false" then
	    last = os.date("%Y%m%d")
	    api:executeString("db insert/call_limit_last_count_of_day/"..key.."/"..last)
	else
        last = api:executeString("db select/call_limit_last_count_of_day/"..key)
    end
	
    local count = 0
	if api:executeString("db exists/call_limit_value_count_of_day/"..key) == "false" then
	    api:executeString("db insert/call_limit_value_count_of_day/"..key.."/0")
	else
        count = tonumber(api:executeString("db select/call_limit_value_count_of_day/"..key)) 	
    end
	
	local current = os.date("%Y%m%d")
	if current ~= last then
	    count = 0
	end
	
	freeswitch.consoleLog("info","Limit the number of calls per day,'"..key.."',limits:"..limits..",count:"..count..".\n")
	if count < limits then
		count = count + 1 
		api:executeString("db insert/call_limit_value_count_of_day/"..key.."/"..count)
		api:executeString("db insert/call_limit_last_count_of_day/"..key.."/"..current)
		stream:write("true")
	else
	    stream:write("false")
    end
elseif pattern == 'count-of-month' then

    local last = nil
	if api:executeString("db exists/call_limit_last_count_of_month/"..key) == "false" then
	    last = os.date("%Y%m")
	    api:executeString("db insert/call_limit_last_count_of_month/"..key.."/"..last)
	else
        last = api:executeString("db select/call_limit_last_count_of_month/"..key)
    end
	
    local count = 0
	if api:executeString("db exists/call_limit_value_count_of_month/"..key) == "false" then
	    api:executeString("db insert/call_limit_value_count_of_month/"..key.."/0")
	else
        count = tonumber(api:executeString("db select/call_limit_value_count_of_month/"..key)) 	
    end
	
	local current = os.date("%Y%m")
	if current ~= last then
	    count = 0
	end
	freeswitch.consoleLog("info","Limit the number of calls per month,'"..key.."',limits:"..limits..",count:"..count..".\n")
	if count < limits then
		count = count + 1 
		api:executeString("db insert/call_limit_value_count_of_month/"..key.."/"..count)
		api:executeString("db insert/call_limit_last_count_of_month/"..key.."/"..current)
		stream:write("true")
	else
        stream:write("false")	
    end
else
    stream:write("false")	
end

File: pyscripts/limit/call_limit.py
import time, datetime
import freeswitch
import re

STM_1 = "09:00:00"
STM_2 = "12:00:00"
STM_3 = "14:00:00"
STM_4 = "18:00:00"
current_hours = time.strftime("%H:%M:%S", time.localtime())
today = time.strftime("%Y-%m-%d", time.localtime())
rootdir = freeswitch.getGlobalVariable("base_dir")
api = freeswitch.API()
file_path = rootdir+"/scripts/pyscripts/sipnumber/number"

def file_list(filename):
    num_list = []
    file = open(filename,"r")
    for line in file.readlines():
        line = line.strip('\n')
        num_list.append(line)
    file.close()
    return num_list

def get_date_time_now():
    today = time.strftime("%Y-%m-%d", time.localtime())
    if api.executeString("db exists/call_limit_last_count_of_day/data_time") == "false":
        api.executeString("db insert/call_limit_last_count_of_day/data_time/"+str(today))
    return api.executeString("db select/call_limit_last_count_of_day/data_time")

def set_data_time_now():
    today = time.strftime("%Y-%m-%d", time.localtime())
    api.executeString("db insert/call_limit_last_count_of_day/data_time/"+str(today))
    return False

def check_gateway_limits_per_hours(key, limit, upgw):
    start_table = []
    new_table = []
    count = 0
    current_hours = time.strftime("%H:%M:%S", time.localtime())
    last_day = get_date_time_now()
    hours_ago = (datetime.datetime.now()+datetime.timedelta(hours=-1)).strftime("%H:%M:%S")
    start_table.append(current_hours)
    if today != last_day:
        for k in range(len(upgw)):
            api.executeString("db delete/call_limit_value_per_hours/"+str(upgw[k]))
    if api.executeString("db exists/call_limit_value_per_hours/"+str(key)) == "false":
        api.executeString("db insert/call_limit_value_per_hours/"+str(key)+"/"+str(start_table))
    else:
        string = api.executeString("db select/call_limit_value_per_hours/"+str(key))
        new_str = string.replace('[','').replace(']','').replace(",","")
        table_list = new_str.split()
        if len(table_list) < int(limit):
            for i in range(len(table_list)):
                if table_list[i] > hours_ago:
                    new_table.append(table_list[i])
            new_table.append(current_hours)
            api.executeString("db insert/call_limit_value_per_hours/"+str(key)+"/"+str(new_table))
        else:
            if len(table_list) == int(limit):
                for i in range(len(table_list)):
                    if table_list[i] > hours_ago:
                        new_table.append(table_list[i])
                if len(new_table) == int(limit):
                    return False
                else:
                    new_table.append(current_hours)
                    api.executeString("db insert/call_limit_value_per_hours/"+str(key)+"/"+str(new_table))
                    return True
    return True
   
def check_dst_number_onece(caller, dst, upgw):
    table = []
    gateway = "gw_"+str(caller)
    table.append(str(dst))
    last_day = get_date_time_now()
    if api.executeString("db exists/call_limit_value_diffnum_onece/"+gateway) == "false":
        api.executeString("db insert/call_limit_value_diffnum_onece/"+gateway+"/"+str(table))
    else:
        if last_day != today:
            for k in range(len(upgw)):
                set_data_time_now()
                api.executeString("db delete/call_limit_value_diffnum_onece/"+"gw_"+str(upgw[k]))
        string = api.executeString("db select/call_limit_value_diffnum_onece/"+gateway)
        new_str = string.replace('[','').replace(']','').replace(",","")
        table_list = new_str.split()
        for i in range(len(table_list)):
            if str(table_list[i]) == str(dst):
                return False
        table_list.append(str(dst))
        api.executeString("db insert/call_limit_value_diffnum_onece/"+gateway+"/"+str(table_list))
    return True

def check_gateway_limits_per_day(key, limits, upgw):
    last_day = get_date_time_now()
    if today != last_day:
        for k in range(len(upgw)):
            api.executeString("db delete/call_limit_value_count_of_day/"+str(upgw[k]))
    if api.executeString("db exists/call_limit_value_count_of_day/"+str(key)) == "false":
        api.executeString("db insert/call_limit_value_count_of_day/"+str(key)+"/1")
    else:
        count = api.executeString("db select/call_limit_value_count_of_day/"+str(key))
        if int(count) < int(limits):
            total = int(count) + 1
            api.executeString("db insert/call_limit_value_count_of_day/"+str(key)+"/"+str(total))
            return True
        else:
            return False
    return True

def db_insert(current_pos):
     api.executeString("db insert/gateway_route/current_pos/"+str(current_pos))

def db_select():
    return api.executeString("db select/gateway_route/current_pos")

def db_exists():
    if api.executeString("db exists/gateway_route/current_pos") == "false":
        api.executeString("db insert/gateway_route/current_pos/0")


def fsapi(session, stream, env, args):
    count, fullgw, current_pos, new_pos, idx, caller, dstnumber = 0, 0, 0, 0, 0, 0, 0
    upgw = []
    if args:
        string = re.split('\s+', args)
    else:
        freeswitch.consoleLog('info', 'must argv profile %s\n' % args)
    argv1 = string[0]
    argv2 = string[1]
    caller = session.getVariable("caller_id_number")
    dstnumber = session.getVariable("destination_number")
    if STM_1 < current_hours < STM_2 or STM_3 < current_hours < STM_4:
        upgw_list = file_list(file_path)
        for gw in range(len(upgw_list)):
            if re.search(r"State\s+REGED\s+",api.executeString("sofia status gateway "+upgw_list[gw])):
                upgw.append(upgw_list[gw])
                count = count + 1
        db_exists()
        current_pos = db_select()
        if int(current_pos) >= count:
            current_pos = 0
        new_pos = int(current_pos) + 1
        idx = int(current_pos)
        db_insert(new_pos)
        info = api.executeString("show channels")
        while True:
            if re.search("sofia/gateway/"+upgw[idx],info) is None:
                if check_gateway_limits_per_day(upgw[idx],argv1,upgw) and check_gateway_limits_per_hours(upgw[idx],argv2) and check_dst_number_onece(upgw[idx],dstnumber):
                    stream.write(str(upgw[idx]))
                    freeswitch.consoleLog("info", "Discovery of available gateway: %s\n" % upgw[idx])
                    break
            if idx == int(current_pos) - 1:
                freeswitch.consoleLog("ERR", "unavailable gateway\n")
                break
            idx = idx + 1
            if int(current_pos) >= count:
                idx = 0
            if idx >= count:
                freeswitch.consoleLog("ERR", "unavailable gateway\n")
                break
    else:
        freeswitch.consoleLog("ERR", "Not in callup time\n")
